pkg/storage/memorystorage/memory_storage.go

```diff
@@ -0,0 +1,94 @@
+package memorystorage
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	internal "github.com/clusterpedia-io/api/clusterpedia"
+	"github.com/clusterpedia-io/clusterpedia/pkg/storage"
+	cache "github.com/clusterpedia-io/clusterpedia/pkg/storage/memorystorage/watchcache"
+	utilwatch "github.com/clusterpedia-io/clusterpedia/pkg/utils/watch"
+)
+
+type StorageFactory struct {
+}
+
+func (s *StorageFactory) NewResourceStorage(config *storage.ResourceStorageConfig) (storage.ResourceStorage, error) {
+	storages.Lock()
+	defer storages.Unlock()
+
+	gvr := schema.GroupVersionResource{
+		Group:    config.GroupResource.Group,
+		Version:  config.StorageVersion.Version,
+		Resource: config.GroupResource.Resource,
+	}
+
+	if resourceStorage, ok := storages.resourceStorages[gvr]; ok {
+		watchCache := resourceStorage.watchCache
+		if config.Namespaced {
+			watchCache.KeyFunc = cache.GetKeyFunc(gvr, config.Namespaced)
+			watchCache.IsNamespaced = true
+		}
+		if _, ok := watchCache.GetStores()[config.Cluster]; !ok {
+			resourceStorage.watchCache.AddIndexer(config.Cluster, nil)
+		}
+
+		resourceStorage.crvSynchro.SetClusterResourceVersion(config.Cluster, "0")
+
+		return resourceStorage, nil
+	}
+
+	watchCache := cache.NewWatchCache(100, gvr, config.Namespaced, config.Cluster)
+	config.WatchCache = watchCache
+	resourceStorage := &ResourceStorage{
+		incoming:      make(chan ClusterWatchEvent, 100),
+		Codec:         config.Codec,
+		crvSynchro:    cache.NewClusterResourceVersionSynchro(config.Cluster),
```

Comment:
Does the initialization of resource storage necessarily need to be bound to a cluster name? I think resource storage is cluster free and then when ready to watch, a watcher is initialized based on the set of clusters passed in, what do you think?

If the resource storage is cluster bound, it would be strange to use a resource storage with a bound cluster to query multiple cluster resources.

Reply:
ClusterResourceVersionSynchro is a struct hold the cluster and the latest resourceVersion of event of this cluster.We encode a  event with the latest resourceVersion of all clusters,so it is necessary.

`storages *resourceStorages` is a struct which hold the all resources stores and latest events in all clusters. The methd of  `NewResourceStorage ` will excute many times in  multiple pediacluster  reconcile logic and finally Initialize `storages *resourceStorages` .The implement logic of NewResourceStorage is also strange  fo us. We have discussed and have no ideas. Hope for any other good suggestions.

Reply:
There is an important concept that resource storage is cluster-independent in APIServer because it needs to support querying an arbitrary set of clusters.

Can we inertly handle cluster information when creating resources or when watching, e.g. inertly initializing indexer `wc.AddIndexer(newCluster, nil)`

Reply:
If there must be some cluster-specific initialization, I think it might be possible to add a PrepareCluster (or some other name) method to the StorageFactory to do some special operations on clusters, and also to correspond to CleanCluster.

Keeping the Resource Storage interface independent of the specific cluster is an important thing

Reply:
It seems that it can't be solved by  inertly way. I prefer adding a PrepareCluster method to the StorageFactory.

Reply:
It seems that it can't be solved by inertly way. I prefer adding a PrepareCluster method to the StorageFactory.

Reply:
```
type StorageFactory interface {
	GetResourceVersions(ctx context.Context, cluster string) (map[schema.GroupVersionResource]map[string]interface{}, error)
	PrepareCluster(cluster string, gvr schema.GroupVersionResource) error
	CleanCluster(ctx context.Context, cluster string) error
```